#2 Unread Message Counter

pranav1416 opened the pull request:
PR with updates that resolves #2 
![PR#2_1](https://user-images.githubusercontent.com/20497587/126868345-6cf59efa-3817-4f10-a4b7-a1c9deb09d5f.png)
![PR#2_2](https://user-images.githubusercontent.com/20497587/126868347-201a3c21-d7cd-4428-85b0-4799f5b64edd.png)

New Changes:
![PR2_1](https://user-images.githubusercontent.com/20497587/126926915-d2e3b4d7-c088-43de-8ae1-351071f31065.png)
![PR2_2](https://user-images.githubusercontent.com/20497587/126926935-bb74312c-527e-46ec-9b7f-40e657e895af.gif)

Commit:
Feature: Unread message counter

Displays unread message notification with count.

Change log:
- Added bug fixes of #1
- Added isRead property to Message model.
- Created ‘/api/messages/read’ route to have post request taking in otherUser id to update read status of messages in the user’s conversation.
- Message is marked read when user clicks on the conversation from the sidebar or clicks on the input field of the active chat from unfocused state of input.
- Added reducer actions in conversation reducer to handle read message and conversations update.

Commit:
Feature: Unread message counter

Displays unread message notification with count.

Change log:
- Added bug fixes of #1
- Added isRead property to Message model.
- Created ‘/api/messages/read’ route to have post request taking in otherUser id to update read status of messages in the user’s conversation.
- Message is marked read when user clicks on the conversation from the sidebar or clicks on the input field of the active chat from unfocused state of input.
- Added reducer actions in conversation reducer to handle read message and conversations update.

Commit:
Feature: Unread message counter

Displays unread message notification with count.

Change log:
- Added bug fixes of #1
- Added isRead property to Message model.
- Created ‘/api/messages/read’ route to have post request taking in otherUser id to update read status of messages in the user’s conversation.
- Message is marked read when user clicks on the conversation from the sidebar or clicks on the input field of the active chat from unfocused state of input.
- Added reducer actions in conversation reducer to handle read message and conversations update.

Commit:
Feature: Unread message counter

Displays unread message notification with count.

Change log:
- Added bug fixes of #1
- Added isRead property to Message model.
- Created ‘/api/messages/read’ route to have post request taking in otherUser id to update read status of messages in the user’s conversation.
- Message is marked read when user clicks on the conversation from the sidebar or clicks on the input field of the active chat from unfocused state of input.
- Added reducer actions in conversation reducer to handle read message and conversations update.

Commit:
Feature: Unread message counter and added security to WebSockets implementation

Changes to Socket security:
- Socket only emits new message to intended user and does not broadcast it to all active sockets.
- Changed onlineUsers type to Object from Array
- Refractor code for fetching onlineUsers in backend routes [server/routes/api/] for users, messages and conversations.

Changes for Feature: unread message counter
- If conversation is clicked or input field in the conversation is clicked or gains focus —> unread messages are marked read.
- Added functionality to handle click and focus in Input component in ActiveChat and ChatContent in Sidebar component.
- Modified Chat component to have state mapped to the props
- Added reducer action `SET_READ_MESSAGES` in conversation reducer, when dispatched it updates the messages (with updated read status) in the store.

Commit:
Feature: Unread message counter and added security to WebSockets implementation

Changes to Socket security:
- Socket only emits new message to intended user and does not broadcast it to all active sockets.
- Changed onlineUsers type to Object from Array
- Refractor code for fetching onlineUsers in backend routes [server/routes/api/] for users, messages and conversations.

Changes for Feature: unread message counter
- If conversation is clicked or input field in the conversation is clicked or gains focus —> unread messages are marked read.
- Added functionality to handle click and focus in Input component in ActiveChat and ChatContent in Sidebar component.
- Modified Chat component to have state mapped to the props
- Added reducer action `SET_READ_MESSAGES` in conversation reducer, when dispatched it updates the messages (with updated read status) in the store.

Commit:
Merge branch 'feature/2-read-status-messages' of https://github.com/pranav1416/messenger-app into feature/2-read-status-messages

Commit:
Bug fix patch

Commit:
Bug fix patch

kuperavv reviewed (commented):
This looks good overall. One thing I didn't see from your UI images is that there should be a small image of the user avatar that appears under the the last read message. Can you add that?

Commit:
Feature: Added read message receipts

Added a feature to display read receipts to the sender, i.e. if the other user has read the message
Change log:
- added socket action `read-message`
- modified Message component to calculate last read message by other user from store and set component state with the message ID
- modified SenderBubble component to display Avatar for read message receipts

Commit:
Feature: Added read message receipts

Added a feature to display read receipts to the sender, i.e. if the other user has read the message
Change log:
- added socket action `read-message`
- modified Message component to calculate last read message by other user from store and set component state with the message ID
- modified SenderBubble component to display Avatar for read message receipts

Commit:
Optional improvements

- Converted to functional components from class components for Home, Input (ActiveChat), Chat, CurrentUser, Search (in Sidebar) in components.
- Added UserMenu component in Sidebar to function as dropdown menu.
- Moved Logout from Home to UserMenu and added functionality.

Commit:
Feature: Added typing status indicator and improvements

- Added TypingIndicator component with transition effects for displaying typing status.
- Used sockets to implement typing status indicator.
- Added reducer action in conversation reducer to update and send typing status.

Commit:
Feature: Added typing status indicator and improvements

- Added TypingIndicator component with transition effects for displaying typing status.
- Used sockets to implement typing status indicator.
- Added reducer action in conversation reducer to update and send typing status.

0916dhkim reviewed (approved):
@pranav1416 Good job addressing Vince's suggestions! This PR is approved :+1: 
Here are things we can improve:
- The position of unread counter is different from the spec.
- There is a bug regarding tracking online user's socket ID.
- Minor cleanups.

See my comments below for more details. You can merge this branch when you think it is ready without additional review.

Commit:
Feature updates for unread messages and conversations

- Repositioned unread count badge to align vertical center of conversation listing
- Repositioned typing status indicator to always appear at bottom of the conversation
- Added scrolling to bottom feature, on overflow and new message, user window scrolled to bottom

Commit:
Feature updates for unread messages and conversations

- Repositioned unread count badge to align vertical center of conversation listing
- Repositioned typing status indicator to always appear at bottom of the conversation
- Added scrolling to bottom feature, on overflow and new message, user window scrolled to bottom

Commit:
Improvements to code

- Removed mapStateToProps and mapDispatchToProps
- Replaced with useSelector and useDispatch to convert the code completely to use hooks

Commit:
Minor update

- Moved clearOnLogout from UserMenu to thunk function

pranav1416 committed:
Merge pull request #9 from pranav1416/feature/4-optional-improvements

#4 Optional improvements

- Converted Class components to functional components
- Converted code to purely use React Hooks
- Added dropdown menu in sidebar
- Moved logout button to the newly added dropdown menu

pranav1416 committed:
Rebase with develop

pranav1416 committed:
Merge branch 'feature/2-read-status-messages' of https://github.com/pranav1416/messenger-app into feature/2-read-status-messages
